### ISSUE TITLE
Rationalize URL-guessing methods in SimpleSAML\Utils\HTTP

### ISSUE THREAD

**Issue by jaimeperez** [simplesamlphp/simplesamlphp#419]
We have a bunch of methods in `SimpleSAML\Utils\HTTP` handling URLs, mostly the current URL. Most of them were buggy because they were trying to use whatever was configured in the `baseurlpath` configuration option, but completely ignoring the path in there, leading to broken URLs. See #396

Additionally, the URL guessing logic was also broken in different setups, mostly when run with _nginx_ #391, #386 and #5.

However, the fixes for these issues led to some other issues with Windows systems or with setups where web server aliases were used together with symbolic links in the file system. All of that was fixed, and led to yet another issue, reported in #418.

The problem is that we have a set of methods that are not fully coherent. They try to use whatever is specified in `baseurlpath`, but not too hard. In general, we should have:
- [ ] Some method to guess the current URL (and the associated methods to remove the query and other parts). This method should **NOT** take the `baseurlpath` configuration option into account, so that it can be used by scripts not belonging to SimpleSAMLphp.
- [ ] Some method to obtain the URL of a script belonging to SimpleSAMLphp, this time, of course honoring `baseurlpath`.
- [ ] Pages should always try to build URLs based on the path to the script, relative to SimpleSAMLphp's installation. This way, we can just get the `baseurlpath` and append the relative path. A good example of this is `SimpleSAML\Module::getModuleURL()`.

This is still a bit challenging, since sometimes we will **really** want to get our current URL to generate a link, but we don't know what's the script in use. An example of this is the language bar, generating links to change the language of the current page. The links in this bar are generated by a common template file, included from every page, and therefore, you never know what's the main script. In cases like this, we can make use of the logic that we have currently, which tries to see if we are running a SimpleSAMLphp script, and then builds the URL mixing the `baseurlpath` together with the path to the script, relative to SimpleSAMLphp's base directory (`/whatever/path/to/ssp/www/`).

We need to rationalize all of this, probably create new functions, rename some existing ones to make their names more intuitive and clear, and review **all the code generating links**.

**Comment by calebdelnayatgfs** [comment]
I'll share some thoughts on this related to reverse proxies:

The 1.14.5 URL change and #418 introduced some problems for us. We we're using SimpleSAMLphp with the Drupal [simplesamlphp_auth module](https://www.drupal.org/project/simplesamlphp_auth). Our Drupal site is hosted behind a reverse proxy which does SSL termination. The actual web server on which Drupal is running does not have SSL setup, so connections from the reverse proxy to the application server are HTTP on port 80. Currently we have some logic inside our `config.php` which checks `$_SERVER` for `HTTP_X_FORWARDED_PROTO` to determine whether the site is behind a reverse proxy and what the original protocol used was. If the value is `https`, then `$_SERVER['HTTPS']` is set to `on`. Over in our `config.php` we set `$config['baseurlpath']` based on the `HTTP_X_FORWARDED_PROTO` header and `$_SERVER['HTTP_HOST']`. WE assume the default port for the protocol, so no port number is ever added to the base URL that we construct.

With the fix in this issue, the use of `\SimpleSAML\Utils\HTTP::getServerPort()` causes SimpleSAMLphp in our environment to produce urls like https://hostname:80/saml/whatever, which doesn't work since we're not running SSL on port 80. This is our fault in a way for not rewriting other global connection state values in `$_SERVER`, like the port. I can work around this issue by overriding `$_SERVER['SERVER_PORT']` in our `config.php`, like we're doing with `$_SERVER['HTTPS']`, but it would be nice if SimpleSAMLphp provided a better mechanism to solve this problem.

Some solutions might be:
1. Look for the `HTTP_X_FORWARDED_PORT` and `HTTP_X_FORWARDED_PROTO` values before checking `SERVER_PORT` and `HTTPS` values.
2. Allow SimpleSAMLphp to be given a state object similar to `$_SERVER`, which would default to `$_SERVER` if not defined, rather than using `$_SERVER` directly.

Note that `HTTP_X_FORWARDED_PROTO` and `HTTP_X_FORWARDED_PORT`, which represents the HTTP request headers `X-Forwarded-Protocol` and `X-Forwarded-Port` respectively, are non-standard but common headers added by reverse proxies. For example, [AWS ELB documentation](http://docs.aws.amazon.com/ElasticLoadBalancing/latest/DeveloperGuide/x-forwarded-headers.html). Wikipedia has a related article, [X-Forwarded-For](https://en.wikipedia.org/wiki/X-Forwarded-For). Searching around a bit more shows there are other headers for the same purpose out in the wild. [RFC-7239](https://datatracker.ietf.org/doc/rfc7239/) attempts to standardize this, though I'm not sure if it is implemented anywhere yet. For those reasons, my second suggested solution would be more flexible.

Note this only applies to cases where `baseurlpath` isn't being used. Since we can control the value of `baseurlpath` programmatically, we can usually set the proper value there, but as @jaimeperez mentions, there are cases where SimpleSAMLphp needs to guess the request URL and currently uses the `$_SERVER` global to do so.

**Comment by jaimeperez** [comment]
Hi @calebdelnayatgfs! Thanks for the feedback!

Let me make a few remarks:

* The `\SimpleSAML\Utils\HTTP::getServerPort()` method did not change in any way in the last versions. In fact, it's been untouched for a long time, mainly due to its simplicity.  The change you are observing in that behaviour is due instead to a change in how you compute URLs, whether you use the `baseurlpath` from the configuration or the environment. Previously, we were assuming **all URLs** were part of SimpleSAMLphp, which is obviously not true for any third-party file (like the Drupal module. It was working in most simple scenarios because people tends to do the same thing you are doing, but URL rewriting is rare. In that case, applying (**partially**) the contents of `baseurlpath` worked, just because the only thing that was changing was the protocol / port or server name, but the path was unchanged. The code was buggy because it was completely ignoring the path in the configuration option, so that _https://example.org/simplesaml/_ as a canonical URL set in the configuration, would be ignored and URLs like _https://example.org/module.php/somemodule/somescript_ would be generated if the environment of the server differs. This was wrong, and it had to be corrected. Unfortunately, this led to a new bug affecting third-party scripts using SSP, like your case.

* The proper way for SSP to work is to honor `baseurlpath` for its own URLs, but not for any other. Therefore, it is the responsibility of the application using SSP to make sure SSP uses the right URLs to refer to the application.

* The use of the `X-Forwarded-*` HTTP headers is broken. And it is because it's not a standard (so implementations differ), and more importantly because there's absolutely no control on the headers. For what it's worth, those headers are user-provided, and therefore cannot be trusted. The security considerations in RFC 7239 make it very clear, and even when they address one issue, they recognize other issues without solution arise. This is the reason why we have denied support for `X-Forwarded-For-*` headers multiple times now, and why the first solution you are proposing is a no-go.

* PHP provides `$_SERVER` and populates environment data in there for a reason. That's the place where a PHP script should look for environment information like the port, whether HTTPS is used or not, etc. Having an alternative, global variable for its exclusive use in SimpleSAMLphp doesn't make much sense, since we already have `$_SERVER`. Besides, global variables are considered bad practice.

* There are two different ways to solve the problem here (for third-party scripts). One, when the problem appears while trying to authenticate the user, we can compute our self URL on our own (taking whatever considerations we need), and pass it to SimpleSAMLphp via the `ReturnTo` parameter that both `requireAuth()` and `login()` accept in the API. The alternative is to modify the contents of `$_SERVER` to reflect the environment as seen by the user, just like you are doing. Note that this always works, regardless of you being able to pass a return URL or not. The key here is to do it **coherently**. If you set `$_SERVER['HTTPS']` to `on` but don't modify the port, you are indeed saying your URLs should be _https://example.com:80_. Therefore, if you need to modify something in `$_SERVER`, you should modify **all the variables related**, such as the port in this case.

I don't know how Drupal deals with this. It might be that they don't even try (or maybe they don't need to, if it can manage fine with relative URLs all the time). However, if there's a mechanism to tell Drupal to use a different canonical URL than the one observed in the environment, the SimpleSAMLphp authentication module should also use that to either pass the appropriate `ReturnTo` parameter to SimpleSAMLphp or populate a new environment in `$_SERVER`, guessing the values from the mechanism provided by Drupal. If it doesn't, it's a bug in the module.

Now, for those deploying something like this Drupal module, without the possibility of modifying its code, there's another alternative to get things right. PHP allows you to prepend a script for every request, so that can be used to modify the environment and make it resemble what your users are seeing. This should of course be a last resort, but it will always work.

**Comment by tdm4** [comment]
@calebdelnayatgfs We had the same problem. I'm not sure what version of simplesamlphp we have (perhaps @jaimeperez could tell me where the version is stored in the copy I have, it was inherited from another development project)

I had to resort to hacking isHTTPS() to:

    public static function isHTTPS() {

                if($_SERVER['HTTPS'] === 'off') {
                   return FALSE;
                } else {
                   return TRUE;
                }
           }

And also had to hack getServerPort() to:

        private static function getServerPort() {

                if (isset($_SERVER["SERVER_PORT"])) {
                        $portnumber = $_SERVER["SERVER_PORT"];
                } else {
                        $portnumber = 80;
                }
                $port = ':' . $portnumber;

                if (self::getServerHTTPS()) {
                        if ($portnumber == '443') $port = '';
                } else {
                        if ($portnumber == '80') $port = '';
                }

                #return $port;
                return ':443';

        }

Because the site is all https and it is SSL Offloaded (SSL is terminated on Amazon ELB).
The $_SERVER['HTTPS'] variable is set by nginx based on the X-Forwarded-Proto header from the load balancer.

How else is nginx going to know the connection is https if it can't supposedly "trust" X-Forwarded-Proto?

**Comment by tdm4** [comment]
I tried setting baseurlpath to https://mysite.example.com/simplesaml/ but that led to a nice broken simplesaml page which gave an exception.

**Comment by jaimeperez** [comment]
Hi @tdm4!

The first thing you should do is to update to the latest version. Plenty of bugs have been fixed since 1.13.2, especially regarding how SimpleSAMLphp builds URLs.

What exception was that? What did the logs say? If your canonical path to SSP is `https://mysite.example.com/simplesaml/` and you specify that in the `baseurlpath` configuration option, SimpleSAMLphp will use that (at least in the last versions, previously it tried to use it but ended up mixing things if URL rewriting was performed).

Regarding `X-Forwarded-Proto` and _nginx_, your web server does **not** need to know if HTTPS is used or not. It's your **application** the one that needs to know, and in this case, SimpleSAMLphp. If you have TLS (I really hope not SSL) offloaded to a reverse proxy or load balancer, the way to tell SimpleSAMLphp to use build URLs using HTTPS is the `baseurlpath` configuration option. As I said previously, the _X-Forwarded-*_ headers do not provide any security and therefore cannot be trusted.

If you also need to build URLs in your own application, you may want to trust those headers **if you are certain that the reverse proxy or load balancer is clearing them** before setting them. In any case, I'd say it's not really worth it. We are in 2016, and you are already using HTTPS, so forget about _X-Forwarded-*_, enforce all your traffic to be through HTTPS, and even enable [_Strict Transport Security_](https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security). Then you can just set `$_SERVER['HTTPS']` to `"on"` and `$_SERVER['SERVER_PORT']` to `443` **unconditionally** in your web server, and forget about the problem.

**Comment by pitbulk** [comment]
Hi @jaimeperez , we are having a similar discussion about the use of X-Forwarded-* at php-saml.

Initially we did not support it, later we added its support on [57](https://github.com/onelogin/php-saml/issues/57) and now @dhensby warn me about its use, and I found that interesting thread.

A couple of days ago, we deactivated it use by default, and added a var to enable it [158](https://github.com/onelogin/php-saml/pull/158).

Now @dhensby offered a new way to control protocol, host and port [152](https://github.com/onelogin/php-saml/pull/152), that overrides the value of $_SERVER.
I have my doubts about that solution and want to know how you guys at simpleSAMLphp think about it.
(maybe you can adopt this solution too if you think is a good idea).


Best regards.

**Comment by jaimeperez** [comment]
¡Hola Sixto! 😄 

TL;DR:
* Stay away from `X-Forwarded-*`, they are evil.
* Provide a way for the user to specify the **full**, **canonical** URL to your library, not only scheme, host and port.
* If you need to _guess_ URLs for the application using your library, don't guess, use the current URL you can observe. If you want to be nice, allow the application to pass you the URL so you don't have to guess, but never try to do the guessing for things that don't belong to your library.

----

Now here are some more details. This has been our biggest source of bugs, trouble, and releases, during the last few months. A complete pain in the ass (pardon my french). The conclusions I just summarized reflect the current status in SimpleSAMLphp, which apparently works fine for everybody, though I don't discard it being broken for someone running some corner-case deployment.

In any case, it all started with someone running SSP behind a reverse proxy doing URL rewriting. They were using `baseurlpath` to specify the canonical URL, but SimpleSAMLphp was buggy as it was taking only the scheme, host and port from that URL, and using the path just as it was in `$_SERVER`. As you can imagine, the result of that was a broken setup when URL rewriting was used to map _https://example.com/foo/bar.php_ in the frontend to _https://example.com/something/else/foo/bar.php_ in the backend, even if `baseurlpath` was configured correctly to be _https://example.com/foo/_.

In order to fix that, I changed the implementation of several methods in our `SimpleSAML\Utils\HTTP` class, breaking on the way support for people using Windows or symlinks.

After fixing it to work again on Windows and when symlinks are used, I totally broke it for people using SimpleSAMLphp as an SP, as it would assume the URL from where authentication was requested was part of SimpleSAMLphp, and the `baseurlpath` was also enforced there, even though it couldn't. The result, again: broken setups.

So here are the reasonings and the lessons learned during the past few months that led to the [current implementation](https://github.com/simplesamlphp/simplesamlphp/blob/master/lib/SimpleSAML/Utils/HTTP.php#L714) we have in SimpleSAMLphp. I hope it helps:

* The `X-Forwarded-*` set of headers is insecure and cannot be trusted. Anyone can spoof them. The only way you can trust them is by clearing them in the reverse proxy or load balancer and setting them again, apart from ensuring your backend servers can only be accessed through the frontend (placing them in a private network, for example). Even the standard admits this weakness, and they suggest no countermeasures at all. Of course, you don't know in your application if the deployment is secure or not, so at the very least you should only allow the use of the `X-Forwarded-*` headers if the user explicitly tells you to (and probably you should also inform your users about this issue, and how to protect themselves). My recommendation is not to use them at all.

* SimpleSAMLphp delegates this to the `baseurlpath` configuration option. That would be equivalent to what @dhensby suggested, but easier to manage, and complete. His proposal (if I read it correctly) lacks the ability to set a _base_ path for your URLs, so reverse proxies doing URL rewriting won't work with that (just as they didn't work with SimpleSAMLphp before we fixed that bug). I personally think it's way easier to ask your users to specify the full, canonical URL where your library can be reached, and forget about it. If you allow setting both an absolute and a relative URL, you can cover the simplest use cases and the most advanced ones.

* Of course, it's nice to have a configuration option you can rely on when you need to build URLs. However, make sure you never use that for URLs pointing to resources that are **not** part of your library. This might seem obvious, but you could find yourself guessing URLs when your library is invoked from external scripts, as I did. Sometimes the difference is a bit blurry. In any case, **make sure you don't guess anything if the current URL is not pointing to any of the scripts in your library**.

* If you want to be super nice to your users, which I guess you want, allow them to pass you the URLs they want to use, so that **they are responsible for their own URLs**, not your library. Alternatively, you can also document that your library will blindly trust the environment for URLs outside it, and if their deployment needs those URLs to be different than what the environment reflects, then they need to change the environment. Meaning: if you have TLS offloaded to a load balancer, you don't have TLS between frontend and backend, but still want URLs to be generated with _https_, make sure to set `$_SERVER['HTTPS']` to `on` before invoking your library. Same applies to host, port, path, and any other data inside `$_SERVER` that would present a mismatch between what you see from the backend, and what's correct from the user's perspective.

I think that's pretty much it. Of course, I can be completely wrong and I'm open to discussion, but that's my experience so far 😉

**Comment by dhensby** [comment]
> SimpleSAMLphp delegates this to the baseurlpath configuration option. That would be equivalent to what @dhensby suggested, but easier to manage, and complete. His proposal (if I read it correctly) lacks the ability to set a base path for your URLs, so reverse proxies doing URL rewriting won't work with that (just as they didn't work with SimpleSAMLphp before we fixed that bug). I personally think it's way easier to ask your users to specify the full, canonical URL where your library can be reached, and forget about it. If you allow setting both an absolute and a relative URL, you can cover the simplest use cases and the most advanced ones.

I agree with this - best to just allow the dev to explicitly set what the URL is, however regarding my PR I'm just trying to get the `php-saml` to work maintaining it's current API - I'm not here to refactor the class or make this kind of decision.

> If you want to be super nice to your users, which I guess you want, allow them to pass you the URLs they want to use, so that they are responsible for their own URLs, not your library. Alternatively, you can also document that your library will blindly trust the environment for URLs outside it, and if their deployment needs those URLs to be different than what the environment reflects, then they need to change the environment. Meaning: if you have TLS offloaded to a load balancer, you don't have TLS between frontend and backend, but still want URLs to be generated with https, make sure to set $_SERVER['HTTPS'] to on before invoking your library. Same applies to host, port, path, and any other data inside $_SERVER that would present a mismatch between what you see from the backend, and what's correct from the user's perspective.

I don't agree with all of this - I don't think expecting users to edit the `$_SERVER` superglobal is an acceptable pattern and I think this should be avoided. `$_SERVER` is global state that should not be edited by the dev.

---

The `X-Forwarded-*` headers are completely unauthenticated and I completely agree they should be ignored as default - this is now implemented in `php-saml` as per a previous PR I submitted that has been merged.

**Comment by pitbulk** [comment]
Thanks for your comments @jaimeperez and @dhensby.

**Comment by rakicivan** [comment]
![image](https://cloud.githubusercontent.com/assets/17475034/25773947/13028978-3287-11e7-9f77-9af55b3205a7.png)
how to fix this? i have SSL

**Comment by tvdijen** [comment]
@jaimeperez do you still have any plans for this, or can we close this issue?

**Comment by porcospino** [comment]
Hello

Sorry to raise this again, but I am in a bit of a bind with this issue.

I have a requirement to use SimpleSAMLphp in an AWS application.  We are building AMIs using packer, and part of the build process installs this library.  These AMIs are used in a Cloudformation template, behind a load balancer.  All instances of this application will use the same IDP, so we can bake the metadata into the AMI.  But we don't know what the URL of the application will be until the stack launches, so I can't specify `baseURL` in the global `$config`.  At launch time, a request to `https://instance.example.com/simplesamlphp` redirects to `https://instance.example.com:80/simplesamlphp`, and we can't retrieve the SP's metatata.

What I'd like to be able to do is specify the port and the scheme in `$config` without having to specify the whole URL in `$baseURL`.  The methods `HTTP::getServerHTTPS()` and `HTTP::getServerPort()` could check for the existence of these configuration variables and use those, rather than guessing them.

**Comment by jaimeperez** [comment]
Hi @porcospino!

If you don't know the URL until the application has been deployed, that means you're just specifying the path to the installation (e.g. `/simplesaml`), and that you are fine with whatever SimpleSAMLphp detects. In that case, remember the configuration file is PHP code, so you can still build a dynamic URL that enforces the use of HTTPS:

```php
   'baseurlpath' => 'https://'.$_SERVER['SERVER_NAME'].'/simplesaml',
```

This is just an example, though. You could also make it depend on environment variables, or whatever you need.
